src/Symfony/Bridge/Twig/Mime/TemplatedEmail.php

```diff
@@ -67,6 +68,21 @@ public function getContext(): array
         return $this->context;
     }
 
+    public function ensureValidity()
+    {
+        if (null === $this->textTemplate && null === $this->htmlTemplate && !$this->getAttachments()) {
```

Comment:
What if you called `text()` instead of `textTemplate()` ? Using a TemplateEmail does not mean that using the non-template setters is not valid (that's a child of Email after all)

Reply:
Thus, by the time `ensureValidatity` is used in the component, the template must have been rendered to populate `text` or `html` already. `generateBody()` won't render the templates by itself. So the existing code is actually right.

Reply:
Maybe that `if` should include all of them?
```
if (null === $this->textTemplate && null === $this->htmlTemplate && !$this->getAttachments()) {
    throw new LogicException('A message must have a text or an HTML part or attachments.');
}
```

Reply:
The point is that those properties are not being populated :/

The code I added to the issue (#39108) is throwing me the `LogicException`: `A message must have a text or an HTML part or attachments`

Reply:
They *are* populated by the `Symfony\Bridge\Twig\Mime\BodyRenderer` (and if you don't wire the body renderer in your mailer, using templates will not work at all)